### DISCUSSION TITLE
"OpenPodcastSync"? New API specification

### DISCUSSION THREAD

**Discussion by thrillfall** [thrillfall/nextcloud-gpodder#91] [Ideas]
The name "gpodder" could also be a bit misleading, since you cannot simply connect clients that only support gpodder (different authentication). How about specifying this as a new API called something like "OpenPodcastSync"? That API specification could then state that:

- gpodder.net supports a subset of the API
- your nextcloud plugin supports the entire API
- there can be different authentication methods in front, but the API itself is the same
- there is a way to discover, for a given server, which capabilities it has

Maybe this could even be a joined work with the people from PodcastIndex (if they agree to keep the API as generic as it is now, without relying on PodcastIndex-specific IDs). In the best case, collaborating with PodcastIndex could convince even proprietary services to implement the protocol, so syncing works seamlessly across different clients.

_Originally posted by @ByteHamster in https://github.com/AntennaPod/AntennaPod/issues/6036#issuecomment-1236896926_

**Comment by keunes** [comment]
Existing & missing API end-points / functionality
---

I fully agree. I had already proposed this to some folks over there and they were open to the idea. James was happy to host/publish agreed standards on https://www.podinfra.net (https://github.com/jamescridland/podinfra.net).

@thrillfall, would you be able to share a (simple) list of calls that gPodder supports? Then we can see where we want/need to expand. In the AntennaPod repo we've already had a discussion somewhere about the calls we'd need, I'll try and dig that up.

**Reply by ByteHamster** [reply]
The gpodder API documentation is here: https://gpoddernet.readthedocs.io/en/latest/api/index.html (we don't use all of the endpoints, only the sync related ones)

**Reply by keunes** [reply]
All right, so I guess these are the main ones we use:
- [Device API](https://gpoddernet.readthedocs.io/en/latest/api/reference/devices.html); to list and update connected devices, but also to get device updates (which overlaps with Episode Actions? This seems a bit odd to me as it's not really related to managing devices - isn't it more logic to have this via the Episode Actions endpoint and have/add a Device filter available there?)
- [Subscriptions API](https://gpoddernet.readthedocs.io/en/latest/api/reference/subscriptions.html); to get & upload subscriptions (& subscription changes)
- [Episode Actions API](https://gpoddernet.readthedocs.io/en/latest/api/reference/events.html); to upload (register) and get download, play, delete, new (really more 'reset') actions.
- [Settings API](https://gpoddernet.readthedocs.io/en/latest/api/reference/settings.html); to get or upload any device (app-wide), subscription and/or episode setting. There are a few 'known' settings (including 'is favourite' strangely enough, I wouldn't consider this a setting) but it seems that any 'setting' would be accepted here.
- [Device Synchronization API](https://gpoddernet.readthedocs.io/en/latest/api/reference/sync.html); to get sync status of devices and start/stop synchronisation. (I would expect this under device management.)

There's also [Podcast Lists API](https://gpoddernet.readthedocs.io/en/latest/api/reference/podcastlists.html), but given that this concerns public lists I don't think we use/need it? (Or is it leveraged for our tags?) And there's [Favorites API](https://gpoddernet.readthedocs.io/en/latest/api/reference/favorites.html), which I also think we don't use, given that it's only to get favourites (not post favourite status).

What I can think of now that AntennaPod is not syncing right now:
* App settings like auto-download (though would be possible via Settings endpoint)
* Subscription settings like playback speed (which I guess could be done via the Settings end-point, though I'm wondering whether it shouldn't be possible to push setting changes via a Subscription Actions endpoint).
* Queue entries & order (incl. 'active'/last played episode)
* Playback history used to build statistics - how is that calculated currently? What data would the server need to have for another device to fully replicate this?

**Reply by ByteHamster** [reply]
The device API has caused a lot of headache in the past because users did not understand that they need to link multiple devices in order to have the same subscriptions on them. I think the nextcloud plugin here does not implement that endpoint, which I think is the best thing to do. I think the use case to have multiple devices where only some synchronize and some don't is something we do not need to support.

> Playback history used to build statistics - how is that calculated currently? What data would the server need to have for another device to fully replicate this?

Each episode has a "time played" and "last played", which is summed up for the statistics screen.

> There's also Podcast Lists API, but given that this concerns public lists I don't think we use/need it?

That's mainly used when using Gpodder as a directory - which does not work well as we can see by the quality of the search results.

**Reply by JonOfUs** [reply]
For me, it sounds most useful to have the following endpoints:
- **Subscription API** - identification using feed/guid, maybe with settings capabilities to set/remove settings values in a hashmap-style environment (Ke-Value). Settings like playback_speed, auto_download would be useful. But just like the Settings API, the settings keys should be completely free to set and we could just maintain lists of known settings in the spec.
- **Episode API** - not Episode Action, since not each action is stored individually (due to overhead) but just one entry per episode. Identification using media url/guid. Maybe useful to have a Key/Value for features here as well? Could contain is_downloaded, is_favorite, ... Alternative would be to have predefined fields for these features, wouldn't be so easy to extend.
- **Settings API** - Global settings in hashmap-style (Key-Value) where clients can use any keys they want. We could maintain a list of known settings keys (like auto-download, theme, ...) in the spec so that other clients could implement them as well if they have similar local settings.
- **Queue API** - Functionality to add/delete items as well as change order. How exactly this will be done is not yet clear.

**Reply by bouktin** [reply]
One functionality that's currently missing is rating a podcast, and rating an episode (1 to 5 stars or whatever).

I like the idea of implementing the feature on the client side or on the server side. A server/service could either publish their own ratings or have users connect their favorite rating app.
The advantage of delegating to an external service is not having to manage fake accounts, fake ratings, public identity & trust.

The ecosystem and open interactions are sorely missing in this case, but here are some hypothetical scenarios:
Client side: 
- User links/signs in within the client app with an external Ratings service account (SensCritique, Spotify, Apple, JustWatch, etc.) 
- Client app intercepts sends ratings API calls to the service

Server side with self publishing:
- Podcast service/server publishes the ratings on its site
- Client app sends ratings API calls to the podcast service/server (alongside other services such as sync)

Server side with external service publishing:
- User links/signs in within the client app with an external Ratings service account (SensCritique, Spotify, Apple, JustWatch, etc.)
- Podcast service/server queues and submits ratings to the Ratings service 
- Client app sends ratings API calls to the podcast service/server (alongside other services such as sync)

**Reply by ByteHamster** [reply]
Ratings are an independent thing. No one would see your ratings if they are just stored on your own Nextcloud.

**Reply by keunes** [reply]
> Ratings are an independent thing. No one would see your ratings if they are just stored on your own Nextcloud.

This depends on the (type of) server: for Nextcloud this wouldn't work (although it could be relevant for major instances like that of a university or family). But for a server like Podfriend (centralised) or Funkwhale (federated) it would surely make sense. So if we can standardise the API calls, I think that'd be great.

**Reply by bouktin** [reply]
Thank you @ByteHamster and @keunes, my suggestion to include ratings (and comments) API is based on these ideas:
* Most podcasts rely on advertising for funding, so ratings and comments are essential for them to develop
* The podcast creators (and industry as a whole) mostly recommends rating or commenting on Apple podcast (as of 2022-09)
* Rating a podcast (and/or commenting, as well as supporting it financially) may become a deciding factor on choosing your podcast client

These make me think worth it to at least discuss ratings/comments as part of the API.

**Reply by thp** [reply]
The thing with especially ratings and comments is if it becomes a "competitive" thing, you have to somehow prevent spamming and so on. Otherwise podcast creators will quickly flood the service with fake users that rate a given podcast quite highly to game the system.

Not having ratings and comments means one doesn't have to spam-filter, moderate or otherwise work with user-created content. In gpodder.net, there is still a "toplist" of episodes and podcasts (so in theory creating users and setting subscriptions can "game" the system), but no way of rating.

Something that might be more useful is the kind of "keyword" rating system I've seen on some sites (airbnb?), like "interview show" (different guests each episode), "dynamic duo" (two recurring moderators), "news show" (irrelevant/outdated after a certain time), "course" (as in educational, might stay relevant even after a few years), ... -- so not just rating on one axis with 5 different values.

Relevant XKCDs: [1098](https://xkcd.com/1098/), [937](https://xkcd.com/937/)

**Reply by bouktin** [reply]
@thp indeed and I'm not suggesting that we should address or solve competitivity. 

I'm suggesting we provide the possibility for the user to rate/comment from their podcast client, preventing them from having to log in to the review/ratings service.

This could encourage external dedicated services to provide interconnection, and let them solve the identity/trust of the reviews and reviewers.

As hinted in my original post:
>  The advantage of delegating to an external service is not having to manage fake accounts, fake ratings, public identity & trust.

**Reply by ByteHamster** [reply]
Shouldn't ratings be an independent API, though? The goal of ratings is that they are public and centralized. The goal of sync would be private and decentralized. So one probably wants to use two different services for these things anyway.

Edit: Also, ratings are not really related to synchronization, and are closely tied to a directory service. So ratings would more fit into an OpenPodcastDirectory API instead.

**Reply by keunes** [reply]
> So one probably wants to use two different services for these things anyway.
> ratings are not really related to synchronization, and are closely tied to a directory service.

I interpreted 'sync' as anything happening between a client and a server – regardless whether that server is a back-end only (gPodder NC app), a full-fletched web-player (Funkwhale/Podfriend) or a directory (can't think of any now).

Whether one wants to separate out rating & other functionality? It depends – I would say what is offered is really at the discretion of the server/service and what is used is at the discretion of the clients/users.

Each client/server should be free to pick and choose the elements of the specs according to their client/server's ability/needs. It's then up to the user to choose the combination that fits their needs.

> So ratings would more fit into an OpenPodcastDirectory API instead.

Having all podcast client-server API stuff together I think is better (more efficient & comprehensible) than fragmenting API specs: for me it doesn't sound logical to (artificially) divide data flows into 2 groups/flows. Worse – have them defined by 2 different groups which cater/engage the same (type of) projects/entities (and would probably largely overlap in practice).

A different end-point? That for sure.

'OpenPodcastSync' was a working title – we can of course decide to change it to OpenPodcastAPI if we feel that fits better :-)

**Reply by JonOfUs** [reply]
I don't think that putting public commenting/rating (with spam control, ...) into the spec is a good idea since this would make implementing it significantly more difficult. And since this spec tries to follow a slightly simpler approach than gpodder.net (without central feed directories, podcast suggestions, ...), such ratings would probably be useless for podcasters since there could be thousands of running sync servers.
> Something that might be more useful is the kind of "keyword" rating system  

I think that we have to follow established standards of podcast rating platforms here. I don't know any of these, but to rate a podcast on different axis won't help podcasters or users outside the own sync server as long as they can't see the ratings in that format and they'll probably only see them on the established platforms.

Rating podcasts + episodes locally and **synchronizing ratings** through the server and maybe even different visibility options for ratings (only me / others on same sync server / public?) sound like a good idea, though. In addition to the ratings, it could be stored for which services the ratings have already been published. This way, ratings of podcasts could be displayed on all synchronized clients and there could even be a server-wide ranking (for family, friends, ... Or on Funkwhale/Podfriend). And ratings would never be published more than once.

**Publishing** ratings is another difficult thing to do. Logging into podcast services through the sync server sounds not very user-friendly as one would have to go to the server control every time to publish a synchronized rating. Since authentication data of these services would be stored inside the server, it would also be necessary to look at the spec even more carefully from a security point of view. Implementing it into podcast clients sounds like a better approach. This way, different podcast clients could provide uploads to different podcast services and could offer to publish all ratings already created at once as soon as a new podcast service is connected.

**Reply by ByteHamster** [reply]
For a first specification of the API, I think we should skip anything directory/rating related. Maybe we want to add that in another version of the API or to another API, but it is not the main use case that we were discussing for OpenPodcastSync. I would say we should get started on the most requested features, not lose ourselves in discussions about features that are way less relevant for now.

**Reply by ByteHamster** [reply]
> Episode API - not Episode Action, since not each action is stored individually (due to overhead) but just one entry per episode. Identification using media url/guid. Maybe useful to have a Key/Value for features here as well? Could contain is_downloaded, is_favorite, ... Alternative would be to have predefined fields for these features, wouldn't be so easy to extend.

I think key/value could be harder to implement server-side. Then you need JOINs and stuff, which could be less efficient than simple table columns when running a larger server.

**Reply by JonOfUs** [reply]
You're probably right, since it's not necessary for episodes, it would definitely perform better without it.
What do you think of the key/value stores in subscriptions and settings? They sound like a useful feature to me. Since there are always way less subscriptions than there are episodes, this would probably even be quite performant.

**Reply by ByteHamster** [reply]
Yeah, I guess we need key/value for settings and subscriptions

**Reply by keunes** [reply]
>  Rating podcasts + episodes locally and **synchronizing ratings** through the server and maybe even different visibility options for ratings (only me / others on same sync server / public?) sound like a good idea
> ...
> **Publishing ratings** is another difficult thing to do.

I'm not sure I understood what exactly mean with the second paragraph, what 'publishing' entails.

In reply to @ByteHamster:
> Maybe we want to add that in another version of the API or to another API, but it is not the main use case that we were discussing for OpenPodcastSync.

For me, synchronisation with Podfriend is one of the reasons why I'm enthusiastic about this initiative. It's a server with which I can sync my my listening progress. And it has ratings. Is it a 'repository'? I wouldn't say so. It'd be cool to not only sync progress but also ratings. And Martin (from Podfriend) said [let's talk](https://podcastindex.social/@martin/105898085171532097).

> I would say we should get started on the most requested features, not lose ourselves in discussions about features that are way less relevant for now.

I know sync is more popular than ratings. But rating episodes/podcasts I feel is something users might be interested in. If not publishing reviews, then certainly consume them to find new podcasts. Just for my own reference:
> small gets audio shows but cannot rate directly each show ([April '22](https://play.google.com/console/u/0/developers/8008695526664634386/app/4974638472012894302/user-feedback/review-details?reviewId=8b499854-0363-46f1-adf2-045dfe46363d&corpus=PUBLIC_REVIEWS))
> great solution. everything works well. I haven't found a rating for podcasts yet. ([April '20](https://play.google.com/console/u/0/developers/8008695526664634386/app/4974638472012894302/user-feedback/review-details?reviewId=cf05c87e-00d3-4e5d-8ac1-df70b1e78e6a&corpus=PUBLIC_REVIEWS))

**Comment by keunes** [comment]
Backwards compatibility / breaking changes
---

Creating a new thread for this comment by @JonOfUs, about using guid as an episode identifier:
> What do you think, how should this change be implemented in gpoddersync?
> Should it just be a v2.0 that is no longer compatible with v1.0?
> This will change a lot, e.g. the unique constraint violation usage probably needs to be reworked if backwards compability is wanted. Or a strict change to GUID as unique index and the deletion of all EpisodeActions without a GUID. But this definitely sounds like a lot of thinking.

Personally I would see backward compatibility as a 'nice to have'. If need be, we should adopt breaking changes: gPodder web is not active any-more and holding back for an inactive (and not super-widely used) service will slow everyone down. Let's focus on active projects like gPodder desktop, kasts, AntennaPod, etc. I have high hopes they'll move along if we engage them.

For example, [podcast guid](https://github.com/Podcastindex-org/podcast-namespace/blob/main/docs/1.0.md#guid) (based on first known URL hash) should probably be adopted (replacing the podcast identifier that has been causing [issues in the past](https://github.com/AntennaPod/AntennaPod/issues/2214)). That'd be a breaking change, too.

**Reply by ByteHamster** [reply]
Publishers will never be consistent about their GUIDs. Not the podcast GUIDs, not the episode GUIDs. Many don't even use episode GUIDs, let alone the podcast ones. So I don't think we should base anything solely on that. I think the current system where we transmit both is the best we can get.

**Reply by JonOfUs** [reply]
To remove gpodder references, an "OpenPodcastSync" implementation would need to be a completely new Nextcloud app as the app ID contains "gpodder". This way, backwards compability with gpoddersync wouldn't be necessary, both sync endpoints could just coexist.

If some podcasts don't provide any episode GUIDs, the current solution to allow episode URL and GUID as identificator at the same time (while GUID superseds URL) would still be necessary. Is that what you mean by the current system?

**Reply by JonOfUs** [reply]
I created a small list of differences from the gpodder.net API, maybe it is better to directly compare this spec to that API since Nextcloud Gpodder is only a subset of it:
1. **No episode actions** but only "time played" and "last played" fields for statistics, leads to smaller database usage
2. **GUID in episode API** (optional), used preferably above media url (less duplicates)
3. **GUID in subscription API?** (optional) also used preferably if available, but way less common, so not that useful
4. **Possibly more flexible authentification** allowing HTTP Auth and maybe even OAuth2 by default? And more auth methods like Nextcloud if implemented by podcast clients
5. **Queue(s) sync**, very useful since many podcast clients have one
6. **No devices API** leading to a less complex design, which should make implementing servers (and using the sync as end-users, as ByteHamster pointed out) much easier
7. **Possibly more flexible settings** with the open Key/Value stores, even on a per-subscription base (gpodder.net also has such a store, but with some official settings keys). Nearly every config could be synced this way. And since the most common settings could be published, this could lead to nice interoperability between podcast clients.

**Reply by thp** [reply]
Not sure if "podcast GUID" is a good idea -- it means one cannot derive the URL from just the GUID alone. Why didn't they just go for storing the "first known URL" (normalized like in the GUID generation, without scheme and trailing slashes) and store that somehow?

This way, it'd be more obvious to a human what a URL could be, and maybe it'd motivate people to [use better URIs](https://www.w3.org/Provider/Style/URI) in the first place.

Supporting the case of "the podcast was at a different URL in the past that we don't control anymore" could be done with the "rel" value of link items, either an existing one, or a new one (`original-podcast-feed-url`? or `podcast-guid-url`): https://microformats.org/wiki/existing-rel-values

The Atom format (which can be mixed with RSS): has `atom:link`, e.g. you could have:

```xml
<atom:link href="www.example.com/podcast.rss" rel="podcast-guid-url" type="application/rss+xml" />
```

(which can be used as-is, as well as deriving a GUID from it, if needed)

**Comment by JonOfUs** [comment]
# Element identification
I think this topic deserves its own thread. It's about how podcasts and episodes are identified across different podcast clients.  
Since no internal IDs can be used, publicly available information about these podcasts have to be used.

gpodder.net identifies subscriptions by feed url and episodes by media url, which lead to many duplicates in the past since both of them can change.  
Episode GUID is quite common, but not each podcast supports it and sometimes (Tagesschau) even guids change.  
Subscription GUID is rarely used, but may be used more in the future, if I understand it correctly.  
Our approach is to make GUID in both cases optional and use URLs for identification if they are missing, but use GUIDs for identification if possible. This approach, however, leads to not very pretty APIs since everywhere where episodes or podcasts are referenced, two separate identification fields have to be available, URL and guid. Episode referencing is, for example, necessary in the queue API.

An alternative approach could be to have a single field combining media URL and guid. But that sounds not very good as well as it does not follow established standards at all.

And in each case, a way of handling detected duplicates would be nice as well since they can still occur. I had the idea of a 'duplicates' endpoint where podcast clients post duplicate episodes (and maybe even subscriptions, if they suddenly 301 their feed). Other clients could then use this information to remove these duplicates from their database (if they haven't already).

**Reply by thp** [reply]
Wanting to have GUID is a nice thing, but obviously (as you wrote above) it's up to the podcast publisher to put sensible values into the feed. If Tagesschau changes the GUID, there's not much one can do apart from trying to heuristically match URL, title, duration, file size, file contents, ... which will lead to other problems and false positives.

The "real" fix there is to fix Tagesschau's feed, not try to come up with a "new standard to work around wrong data". Or to take a quote attributed to [Charles Babbage](https://de.wikipedia.org/wiki/Charles_Babbage):

> On two occasions I have been asked, 'Pray, Mr. Babbage, if you put into the machine wrong figures, will the right answers come out?' I am not able rightly to apprehend the kind of confusion of ideas that could provoke such a question.

Then again, you can't force all publishers to Do The Right Thing, and being able to deal with "non-optimal" feeds makes the sync protocol work for a greater set of podcasts. Podcasts are (thankfully still) decentralized, if you put them all behind a walled garden (*cough* *cough* Spotify) you could of course enforce lots of things.

Every feed needs a URL, and every episode also needs a download URL, which is why we went for those features in the protocol. How stable those URLs are, and if there's also such a thing as GUIDs, is a different topic (and apparently a source of much pain).

---

Stefan created a "best practice" guide for feeds in the past:
https://github.com/gpodder/podcast-feed-best-practice/blob/master/podcast-feed-best-practice.md

Apple also has a "feed requirements" documents for Podcasts:
https://podcasters.apple.com/support/823-podcast-requirements

Also this, "A Podcaster's Guide to RSS":
https://help.apple.com/itc/podcasts_connect/

In general, the "podcasts should have feature X" is wishful thinking, and while there's some chance of trying to lobby publishers into adopting sane defaults, there will always be the odd one out and then it's a question of can the system deal with it gracefully or at least acceptably (having duplicate episodes is annoying, but not as bad as "missing" episodes or the inability to parse feeds in general).

Also, trying to invent "The New Standard" is [XKCD 927](https://xkcd.com/927/) -- better utilize existing things as much as possible and interoperate with what's out there.

You can always send as much data as you have about the episode (e.g. send the media URL, but also send the GUID if available, and let the server side deal with it in the best way possible). This allows the server side to improve in the future with matching/detecting without having to change the protocol or the clients. It also allows "simple" implementations of the server side to just ignore one or the other while still providing interoperability (similarly, the protocol could specify that URL + GUID are "recommended" for server->client messaging, but clients should be able to deal with only one of them being available).

For example:

```json
{
    "guid": "b8de566f-04f9-4efa-b594-f5040d40f990",
    "url": "https://example.com/file.mp3"
}
```

But also allowed:

```json
{
    "guid": null,
    "url": "https://example.net/podcast.ogg"
}
```

This way, turning a file URL into a GUID is up to the receiver, and ideally both would be stored, and GUID would be preferred over the file URL. Then, again, it's an implementation detail what happens if:

* GUID missing, URL matches
* GUID exists and matches, URL matches (very likely the same episode)
* GUID exists and matches, URL differs (probably fine, just different CDN URL?)
* GUID exists and differs, URL matches (ouch?)
* GUID exists and differs, URL differs (very likely a different episode)

**Reply by JonOfUs** [reply]
Yes, that sounds like the best way we can handle episode identification. Nextcloud Gpodder handles episode identification this way and I haven't had any issues with it up to now (haven't tested it extensively, however).  
In Apple's guide it sounds like they try to actually enforce episode GUIDs. I don't know any statistics, but from my knowledge most podcast feeds provide them. So using them (optionally) sounds logical to me.

> GUID exists and differs, URL matches (ouch?)

Tagesschau case here, I think. It's probably best to handle it as **different episodes**. We could put advanced duplicate detection into this (AntennaPod does this), but different implementations of the spec would surely decide differently sometimes which could lead to unexpected behaviour. The only way I see to conveniently prevent/remove duplicates is an actual endpoint for it (in some kind).

**Reply by ByteHamster** [reply]
> GUID exists and differs, URL differs (very likely a different episode)

Very often it is still the same episode. When podcasters replace their media file (fix missing parts etc) and don't know what they are doing, they often also replace the GUID. AntennaPod does [quite a bit of guessing](https://github.com/AntennaPod/AntennaPod/blob/413df734953dbfbaffac312745f7a467c3ec41ff/core/src/main/java/de/danoeh/antennapod/core/storage/FeedItemDuplicateGuesser.java) to detect these.
